### PR TITLE
feat: refactor e2e partition planning algorithm

### PR DESCRIPTION
**Pull request by weeco**
## Summary

This PR completely refactors the e2e probe topic partition planning algorithm to fix critical issues with unnecessary reassignments and improve the overall reliability of partition leadership and replica distribution optimization.

## Key Changes

- **New 3-phase partition planner**: Systematic approach with phase 1 (RF/rack fixes), phase 2 (leader coverage), phase 3 (partition count)
- **Enhanced replica selection**: Rack-aware selector with load balancing and deterministic broker selection
- **Improved logging**: Clear distinction between routine checks ("optimal") vs actual optimization plans
- **Comprehensive test suite**: 13 test cases covering edge cases, deterministic behavior, and the specific bug scenario
- **Deterministic iteration**: Sorted partition/broker IDs to ensure consistent planning results

### Architecture Improvements  
- Separated concerns into distinct phases for better maintainability
- Added `LoadTracker` for intelligent replica placement decisions  
- Introduced `ReplicaSelector` interface for pluggable selection strategies
- Enhanced `Partition` struct to track both actual and preferred leaders

### Logging Enhancements
- No-op scenarios: `"e2e probe topic partition leadership and replica distribution check completed - optimal"`  
- Optimization scenarios: `"plan to change partition leadership and replica placements on e2e topic has been prepared"`

### CODE REVIEW

File: e2e/partition_planner.go [hunk 1]
```diff
@@ -0,0 +1,938 @@
...
+type PlanBuilder struct {
+	state   ClusterState
+	desired Desired
+	tracker *LoadTracker
+
+	// view is our predictive map: partitionID -> replicas (preferred leader at idx 0)
+	view map[int32][]int32
+
+	reassign []Reassignment     // staged reassignments for existing partitions
+	creates  []CreateAssignment // staged creations of new partitions
```
Comment (sbocinec) [inline e2e/partition_planner.go:126]: nit/suggestion: I'd prefer both struct fields to either use singular or plural form to stay consistent, but not mixed

**Review by sbocinec — APPROVED** [review]
Huge effort 🙌  I added a few nit comments, LGTM

File: e2e/partition_planner.go (hunk 1 above)
Comment (weeco) [reply e2e/partition_planner.go]: Fixed!
